pkg/kube/krt/nestedjoinmerge.go

```diff
@@ -0,0 +1,365 @@
+			case controllers.EventAdd:
+				// When a collection is added, subscribe to its events
+				reg := obj.RegisterBatch(subscriptionFunc, true)
+				j.mu.Lock()
+				j.regs[obj.(internalCollection[T]).uid()] = reg
+				j.mu.Unlock()
+			case controllers.EventUpdate:
+				j.handleCollectionUpdate(ev)
+			case controllers.EventDelete:
+				j.handleCollectionDelete(ev)
+			}
+		}
+	}, false)
+	initialCollections := j.collections.List()
+	// Finally, async wait for the primary to be synced. Once it has, we know it has enqueued the initial state.
+	// After this, we can run our queue.
+	// The queue will process the initial state and mark ourselves as synced (from the NewWithSync callback)
+	go j.runQueue(initialCollections, subscriptionFunc, reg)
+
+	return j
+}
+
+func (j *nestedjoinmerge[T]) runQueue(initialCollections []Collection[T], subscriptionFunc func([]Event[T]), reg HandlerRegistration) {
+	// Wait for the container of collections to be synced before we start processing events.
+	if !j.collections.WaitUntilSynced(j.stop) {
+		return
+	}
+	j.mu.Lock()
+
+	// Now that we've subscribed, process the current set of collections.
+	for _, c := range initialCollections {
+		// Save these registrations so we can unsubscribe later if the collection is deleted.
+		// Ensure each sub-collection is synced before we're marked as synced.
+		j.regs[c.(internalCollection[T]).uid()] = c.RegisterBatch(subscriptionFunc, true)
+	}
+
+	regs := append([]HandlerRegistration{reg}, maps.Values(j.regs)...)
+	j.mu.Unlock()
+
+	syncers := slices.Map(regs, func(r HandlerRegistration) cache.InformerSynced {
+		return r.HasSynced
+	})
+
+	if !kube.WaitForCacheSync(j.collectionName, j.stop, syncers...) {
+		return
+	}
+	j.queue.Run(j.stop)
+}
+
+func (j *nestedjoinmerge[T]) handleCollectionUpdate(e Event[Collection[T]]) {
+	innerCollection := e.Latest().(internalCollection[T])
+	log.Debugf("NestedJoinWithMergeCollection: Collection %s (uid %s) updated, recalculating merged values", innerCollection.name(), innerCollection.uid())
+	// Get all of the elements in the old collection
+	oldCollectionValue := *e.Old
+	newCollectionValue := *e.New
+	// Wait for the new collection to be synced before we process the update.
+	if !newCollectionValue.WaitUntilSynced(j.stop) {
+		log.Warnf("NestedJoinWithMergeCollection: Collection %s not synced, skipping update event", newCollectionValue.(internalCollection[T]).uid())
+	}
+	// Stop the world and update our outputs with new state for everything in the collection.
+	j.mu.Lock()
+	defer j.mu.Unlock()
+
+	oldItems := oldCollectionValue.List()
+	// Convert it to a map for easy lookup
+	oldItemsMap := make(map[Key[T]]T, len(oldItems))
+	for _, i := range oldItems {
+		key := getTypedKey(i)
+		oldItemsMap[key] = i
+	}
+	// Now loop through the new collection and compare it to the old one
+	seen := sets.NewWithLength[string](len(oldItems))
+	finalEvents := make([]Event[T], 0, len(oldItems))
+	for _, i := range newCollectionValue.List() {
```

Comment:
when we call these `.List()` on the underlying new/old collections, those collections aren't locked so we could theoretically miss events, right? Or do we register a handler before doing the List, so whatever was missed after the `List()` to get the initial state of the new collection will get updated individually later?

Reply:
Yeah collection updates are a weird case; the Key for collections is their UID, so an update would by definition have to preserve that key I think. Since the key/uid is the same across an update, that would also mean handlers are preserved, so we'd get events that happen after this List call during the regular event handler process. We also ensure it happens after this update func exists because we hold a lock in both